src/cls/rgw/cls_rgw.cc

```diff
@@ -133,6 +149,41 @@ static void get_index_ver_key(cls_method_context_t hctx, uint64_t index_ver, str
   *key = buf;
 }
 
+static void bi_reshard_log_prefix(string& key)
+{
+  key = BI_PREFIX_CHAR;
+  key.append(bucket_index_prefixes[BI_BUCKET_RESHARD_LOG_INDEX]);
+}
+
+// 0x802001_gen_idx
+static void bi_reshard_log_key(cls_method_context_t hctx, uint64_t gen, string& key, const string& idx)
+{
+  bi_reshard_log_prefix(key);
+  char buf[32];
+  snprintf(buf, sizeof(buf), "%llu_", (unsigned long long)gen);
+  key.append(buf);
+  key.append(idx);
+}
+
+static int reshard_log_index_operation(cls_method_context_t hctx, const string& idx,
+                                       uint64_t gen, const cls_rgw_obj_key& key,
+                                       bufferlist* log_bl)
+{
+  string reshard_log_idx;
+  assert(gen >= 1);
```

Comment:
Do we want asserts in cls code? Will this crash the OSD if it fails?

Reply:
well this is C-assert, which means it will abort in debug builds only, iirc;  which is, I think, the correct semantics if we're going to have asserts in the code at all (e.g., we'll have contracts in a couple years, apparently.  I think C assert is ok, I don't like ceph_assert().

Reply:
> Do we want asserts in cls code? Will this crash the OSD if it fails?

Yep, this will crash the OSD. I added this for testing that `gen` can be set correctly, but forgot to remove it before pr.

Reply:
> well this is C-assert, which means it will abort in debug builds only, iirc; which is, I think, the correct semantics if we're going to have asserts in the code at all (e.g., we'll have contracts in a couple years, apparently). I think C assert is ok, I don't like ceph_assert().

Sorry, I don't know the difference between assert() and ceph_assert() before your detailed comment. But I do add the assert() in testing(compiled in DEBUG) to guarantee `gen` is used properly, and plan to remove it before pr. 
I forgot to remove it, otherwise, release builded clusters will be deployed usually, then there is no need to remove assert(). So for this  circumstance, I also think `assert()` is better.